### PR TITLE
DST-193 POC webvowl visualisation

### PR DESCRIPTION
**Pull request by lazcool**

### CODE REVIEW

File: DFC.ServiceTaxonomy.GraphVisualiser/Services/Neo4JToOwlGeneratorService1.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
+﻿using System.Collections.Generic;
```
Comment (lazcool) [inline DFC.ServiceTaxonomy.GraphVisualiser/Services/Neo4JToOwlGeneratorService1.cs]: Looks like the filename has an extra '1' on the end.

File: DFC.ServiceTaxonomy.GraphVisualiser/Queries/GetNodesCypherQuery.cs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        public Query Query
+        {
+            get
+            {
+                // $"match (n:ncs__JobProfile {{uri:\"{fetch}\"}})-[r]-(d) optional match (d)-[r1:esco__relatedEssentialSkill|:esco__relatedOptionalSkill|:ncs__hasUniversityLink|:ncs__hasUniversityRequirement]-(d1) return n, d, r, r1, d1";
+
+                var cypherQuery = new StringBuilder();
+
+                cypherQuery.Append($"MATCH (s {{{MatchPropertyName}:'{MatchPropertyValue}'}})");
+                cypherQuery.Append($"OPTIONAL MATCH (s {{{MatchPropertyName}:'{MatchPropertyValue}'}})-[r1]-(d1)");
```
Comment (lazcool) [inline DFC.ServiceTaxonomy.GraphVisualiser/Queries/GetNodesCypherQuery.cs]: Can we remove the property match here, i.e. (`{{{MatchPropertyName}:'{MatchPropertyValue}'}}`)?

File: DFC.ServiceTaxonomy.GraphVisualiser/Queries/GetNodesCypherQuery.cs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+            if (relationship != null)
+            {
+                Relationships.Add(relationship);
+                result += $" - [{relationship.Type}";
+            }
+            if (destNode != null)
+            {
+                Nodes[destNode.Id] = destNode;
+                result += $" - " + (destNode.Properties.ContainsKey(DestPropertyName) ? destNode.Properties[DestPropertyName].ToString() : string.Empty);
```
Comment (lazcool) [inline DFC.ServiceTaxonomy.GraphVisualiser/Queries/GetNodesCypherQuery.cs]: Shouldn't need the `$` on the string.

File: DFC.ServiceTaxonomy.GraphVisualiser/Services/Neo4JToOwlGeneratorService1.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+            TransformData(nodes, prefLabel);
+            TransformData(relationships);
+
+            var result = new OwlDataModel
+            {
+                Namespace = CreateNamespaces(),
+                Header = CreateHeader(),
+                Settings = CreateSettings(),
+                Class = nodeDataModels.Select(n => CreateClass(n, $"{selectedNodeId - minNodeId}")).ToList(),
+                ClassAttribute = nodeDataModels.Select(n => CreateClassAttribute(n)).ToList(),
```
Comment (lazcool) [inline DFC.ServiceTaxonomy.GraphVisualiser/Services/Neo4JToOwlGeneratorService1.cs]: Should just be able to use the method group on these 3 lines, e.g. `Select(CreateClassAttribute)`

File: DFC.ServiceTaxonomy.GraphVisualiser/Services/OrchardToOwlGeneratorService.cs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+            TransformData(selectedContentTypeDefinition);
+
+            var result = new OwlDataModel
+            {
+                Namespace = CreateNamespaces(),
+                Header = CreateHeader(),
+                Settings = CreateSettings(),
+                Class = nodeDataModels.Select(n => CreateClass(n, selectedContentTypeDefinition.Name)).ToList(),
+                ClassAttribute = nodeDataModels.Select(n => CreateClassAttribute(n)).ToList(),
```
Comment (lazcool) [inline DFC.ServiceTaxonomy.GraphVisualiser/Services/OrchardToOwlGeneratorService.cs]: Again, could use method groups here.

File: DFC.ServiceTaxonomy.GraphVisualiser/Views/Visualise/Viewer.cshtml [hunk 6]
```diff
@@ -0,0 +1,558 @@
...
+                </div>
+            </div>
+
+
+        </div>
+    </main>
+    <aside id="class-properties-popup" class="avgrund-popup">
+        <h2 id="class-properties-popup-label">Class label</h2>
+        <p id="class-properties-popup-properties">
+            properties here
```
Comment (lazcool) [inline DFC.ServiceTaxonomy.GraphVisualiser/Views/Visualise/Viewer.cshtml]: Looks like we didn't end up using avgrund, so we can remove this and its js / css files.

File: DFC.ServiceTaxonomy.GraphVisualiser/Services/Neo4JToOwlGeneratorService1.cs (hunk 1 above)
Comment (IanCrisp) [reply DFC.ServiceTaxonomy.GraphVisualiser/Services/Neo4JToOwlGeneratorService1.cs]: done

File: DFC.ServiceTaxonomy.GraphVisualiser/Queries/GetNodesCypherQuery.cs (hunk 3 above)
Comment (IanCrisp) [reply DFC.ServiceTaxonomy.GraphVisualiser/Queries/GetNodesCypherQuery.cs]: done

File: DFC.ServiceTaxonomy.GraphVisualiser/Controllers/VisualiseController.cs [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+            return Content(owlResponseString, MediaTypeNames.Application.Json);
+        }
+
+        public async Task<ActionResult> GetData(string uri)
+        {
+            const string prefLabel = "skos__prefLabel";
+            var query = new GetNodesCypherQuery(nameof(uri), uri, prefLabel, prefLabel);
+
+            _ = await _neoGraphDatabase.Run(query);
```
Comment (tomtollafield) [inline DFC.ServiceTaxonomy.GraphVisualiser/Controllers/VisualiseController.cs]: What is _? This method doesn't seem to do anything with the result

File: DFC.ServiceTaxonomy.GraphVisualiser/Services/Neo4JToOwlGeneratorService1.cs (hunk 4 above)
Comment (IanCrisp) [reply DFC.ServiceTaxonomy.GraphVisualiser/Services/Neo4JToOwlGeneratorService1.cs]: done

File: DFC.ServiceTaxonomy.GraphVisualiser/Queries/GetNodesCypherQuery.cs [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+            SelectedNodeId = sourceNode.Id;
+
+            string sourceLabel = sourceNode.Properties.ContainsKey(SourcePropertyName) ? sourceNode.Properties[SourcePropertyName].ToString() : string.Empty;
+            string result = $"{sourceLabel}";
+
+            if (relationship != null)
+            {
+                Relationships.Add(relationship);
+                result += $" - [{relationship.Type}";
```
Comment (tomtollafield) [inline DFC.ServiceTaxonomy.GraphVisualiser/Queries/GetNodesCypherQuery.cs]: Is this missing a closing square bracket e.g. [{relationship.Type}]

File: DFC.ServiceTaxonomy.GraphVisualiser/Services/OrchardToOwlGeneratorService.cs (hunk 5 above)
Comment (IanCrisp) [reply DFC.ServiceTaxonomy.GraphVisualiser/Services/OrchardToOwlGeneratorService.cs]: done

File: DFC.ServiceTaxonomy.GraphVisualiser/Views/Visualise/Viewer.cshtml (hunk 6 above)
Comment (IanCrisp) [reply DFC.ServiceTaxonomy.GraphVisualiser/Views/Visualise/Viewer.cshtml]: ok

File: DFC.ServiceTaxonomy.GraphVisualiser/Controllers/VisualiseController.cs (hunk 7 above)
Comment (IanCrisp) [reply DFC.ServiceTaxonomy.GraphVisualiser/Controllers/VisualiseController.cs]: _ is a discard, yes it doesn't use the records

File: DFC.ServiceTaxonomy.GraphVisualiser/Queries/GetNodesCypherQuery.cs (hunk 8 above)
Comment (IanCrisp) [reply DFC.ServiceTaxonomy.GraphVisualiser/Queries/GetNodesCypherQuery.cs]: yes it is for information only - will add it to the string

File: DFC.ServiceTaxonomy.GraphVisualiser/Services/OwlDataGeneratorService.cs [hunk 9]
```diff
@@ -0,0 +1,200 @@
+﻿using System.Collections.Generic;
+using DFC.ServiceTaxonomy.GraphVisualiser.Models;
+using DFC.ServiceTaxonomy.GraphVisualiser.Models.Owl;
+
+namespace DFC.ServiceTaxonomy.GraphVisualiser.Services
+{
+    public abstract class OwlDataGeneratorService
+    {
+        private const string DefaultLanguage = "en";
```
Comment (tomtollafield) [inline DFC.ServiceTaxonomy.GraphVisualiser/Services/OwlDataGeneratorService.cs]: Could these be moved to AppSettings?

File: DFC.ServiceTaxonomy.GraphVisualiser/Views/Visualise/Viewer.cshtml (hunk 6 above)
Comment (IanCrisp) [reply DFC.ServiceTaxonomy.GraphVisualiser/Views/Visualise/Viewer.cshtml]: removed

File: DFC.ServiceTaxonomy.GraphVisualiser/Services/OwlDataGeneratorService.cs [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+        private const string NamespaceName = "NCSnamespace";
+        private const string NamespaceIri = "https://nationalcareers.service.gov.uk";
+        private const string HeaderIri = "https://nationalcareers.service.gov.uk/test/";
+        private const string HeaderAuthor = "National Careers Service";
+        private const string HeaderVersion = "0.0.1(alpha)";
+        private const string DescriptionLabel = "NationalCareersService-ServiceTaxonomy";
```
Comment (tomtollafield) [inline DFC.ServiceTaxonomy.GraphVisualiser/Services/OwlDataGeneratorService.cs]: Could these be moved to App Settings?

Comment (IanCrisp) [reply DFC.ServiceTaxonomy.GraphVisualiser/Services/OwlDataGeneratorService.cs]: yes I can do that

File: DFC.ServiceTaxonomy.GraphVisualiser/Queries/GetNodesCypherQuery.cs (hunk 2 above)
Comment (IanCrisp) [reply DFC.ServiceTaxonomy.GraphVisualiser/Queries/GetNodesCypherQuery.cs]: Amended as suggested and it works

**Pull request merged**